network/p2p/dns/resolver.go

```diff
@@ -109,22 +151,82 @@ func (r *Resolver) LookupTXT(ctx context.Context, txt string) ([]string, error)
 
 // lookupIPAddr encapsulates the logic of resolving a txt through cache.
 func (r *Resolver) lookupTXT(ctx context.Context, txt string) ([]string, error) {
-	if addr, exists, invalidated := r.c.resolveTXTCache(txt); exists {
-		// resolving address from cache
-		r.collector.OnDNSCacheHit()
-		return addr, nil
-	} else if invalidated {
-		r.collector.OnDNSCacheInvalidated()
+	addr, exists, fresh := r.c.resolveTXTCache(txt)
+
+	if !exists {
+		r.collector.OnDNSCacheMiss()
+		return r.lookupResolverForTXTAddr(ctx, txt)
+	}
+
+	if !fresh && r.shouldResolveTXT(txt) {
+		r.unit.Launch(func() {
+			defer r.doneResolvingTXT(txt)
+			_, err := r.lookupResolverForTXTAddr(ctx, txt)
+			if err != nil {
+				// invalidates cached entry when hits error on resolving.
+				invalidated := r.c.invalidateTXTCacheEntry(txt)
+				if invalidated {
+					r.collector.OnDNSCacheInvalidated()
+				}
+			}
+		})
+
 	}
 
-	// resolves txt through underlying resolver
-	r.collector.OnDNSCacheMiss()
+	r.collector.OnDNSCacheHit()
+	return addr, nil
+}
+
+// lookupResolverForIPAddr queries the underlying resolver for the domain and updates the cache if query is successful.
+func (r *Resolver) lookupResolverForTXTAddr(ctx context.Context, txt string) ([]string, error) {
 	addr, err := r.res.LookupTXT(ctx, txt)
 	if err != nil {
 		return nil, err
 	}
 
 	r.c.updateTXTCache(txt, addr) // updates cache
 
-	return addr, err
+	return addr, nil
+}
+
+// shouldResolveIP returns true if there is no other concurrent attempt ongoing for resolving the domain.
+func (r *Resolver) shouldResolveIP(domain string) bool {
+	r.Lock()
+	defer r.Unlock()
+
+	if _, ok := r.processingIPs[domain]; !ok {
+		r.processingIPs[domain] = struct{}{}
+		return true
+	}
+
+	return false
+}
+
+// doneResolvingIP cleans up tracking an ongoing concurrent attempt for resolving domain.
+func (r *Resolver) doneResolvingIP(domain string) {
+	r.Lock()
+	defer r.Unlock()
+
+	delete(r.processingIPs, domain)
+}
+
+// doneResolvingIP cleans up tracking an ongoing concurrent attempt for resolving txt.
+func (r *Resolver) doneResolvingTXT(txt string) {
+	r.Lock()
+	defer r.Unlock()
+
+	delete(r.processingTXTs, txt)
+}
+
```

Comment:
Vishal's [comment](https://github.com/onflow/flow-go/pull/1216/files#r708767925) may also apply here.

Reply:
Removing locks here causes a data race between `shouldResolveIP` and `doneResolvingIP`. A similar data race is also prone for `shouldResolveTXT` and `doneResolvingTXT`.